domain/blockdag/blocknode.go

```diff
@@ -235,3 +266,111 @@ func (node blockNode) String() string {
 func (node *blockNode) time() mstime.Time {
 	return mstime.UnixMilliseconds(node.timestamp)
 }
+
+func (node *blockNode) blockAtDepth(depth uint64) *blockNode {
+	if node.blueScore <= depth { // to prevent overflow of requiredBlueScore
+		depth = node.blueScore
+	}
+
+	current := node
+	requiredBlueScore := node.blueScore - depth
+
+	for current.blueScore >= requiredBlueScore {
+		if current.isGenesis() {
+			return current
+		}
+		current = current.selectedParent
+	}
+
+	return current
+}
+
+func (node *blockNode) finalityPoint() *blockNode {
+	return node.blockAtDepth(node.dag.FinalityInterval())
+}
+
+func (node *blockNode) hasFinalityPointInOthersSelectedChain(other *blockNode) (bool, error) {
+	finalityPoint := node.finalityPoint()
+	return node.dag.isInSelectedParentChainOf(finalityPoint, other)
+}
+
+func (node *blockNode) nonFinalityViolatingBlues() (blockSet, error) {
+	nonFinalityViolatingBlues := newBlockSet()
+
+	for _, blueNode := range node.blues {
+		notViolatingFinality, err := node.hasFinalityPointInOthersSelectedChain(blueNode)
+		if err != nil {
+			return nil, err
+		}
+		if notViolatingFinality {
+			nonFinalityViolatingBlues.add(blueNode)
+		}
+	}
+
+	return nonFinalityViolatingBlues, nil
+}
+
+func (node *blockNode) checkBoundedMergeDepth() error {
+	nonFinalityViolatingBlues, err := node.nonFinalityViolatingBlues()
+	if err != nil {
+		return err
+	}
+
+	finalityPoint := node.finalityPoint()
+	for _, red := range node.reds {
+		doesRedHaveFinalityPointInPast, err := node.dag.isInPast(finalityPoint, red)
+		if err != nil {
+			return err
+		}
+
+		isRedInPastOfAnyNonFinalityViolatingBlue, err := node.dag.isInPastOfAny(red, nonFinalityViolatingBlues)
+		if err != nil {
+			return err
+		}
+
+		if !doesRedHaveFinalityPointInPast && !isRedInPastOfAnyNonFinalityViolatingBlue {
+			return ruleError(ErrViolatingObjectiveFinality, "block is violating objective finality")
```

Comment:
Is "objective finality" the term we're using after all?

Reply:
Nope, changed to "Bounded merge depth" :stuck_out_tongue: